### PR TITLE
Jna kotlin minimal example

### PR DESCRIPTION
**Pull request by jcrist1**
This starts to address [#144](https://github.com/rust-diplomat/diplomat/issues/144). 

I've got a simple minimal version that supports

- [x] project structure
- [x] primitive types
- [x] opaque types
- [x] return owned opaque types
- [x] immutable self borrows on opaque type methods

I'd like to solicit feedback, before going further (also to make the review load more manageable). Still missing:
- [x] ~domain~ group is an ~attr~ is configurable 
- [ ] `lib_name` comes from Cargo.toml
- [x] opaque type borrows ~in non-self other positions~
- [x] structs
- [x] enums
- [x] writeable
- [x] slices
  - [x] primitive slices
  - [x] str slices
  - [x] owned slices should be kotlin arrays
  - [x] slices of strings
- [x] strings
- [x] borrows
- [ ] concurrency protection (which I would like for mutable access)
- [ ] update documentation, examples, tests, and code comments
  - [ ] documentation
  - [ ] examples
  - [x] feature_tests
  - [ ] code comments
- [x] nullables
- [ ] fallibles
- [ ] iterables

### Testing
I created a simple test project with the following source 
```rust
#[diplomat::bridge]
mod ffi {
    #[diplomat::opaque]
    struct SomeStruct {
        s: String,
    }

    impl SomeStruct {
        pub fn create(in1: i32, in2: i32) -> Box<SomeStruct> {
            Box::new(SomeStruct {
                s: format!("{}", in1 + in2),
            })
        }

        pub fn do_something(&self, in1: i32) -> i32 {
            self.s.len() as i32 + in1
        }
    }
}
```
with the following `Cargo.toml`
```toml
[package]
name = "somelib"
version = "0.1.0"
edition = "2021"

[lib]
crate_type = ["cdylib"]
name = "somelib"

[dependencies]
diplomat = {path = "../diplomat/macro/"}
diplomat-runtime = {path = "../diplomat/runtime"}
```
Running
```sh
mkdir -p ~/tmp/test_kt
cargo build --release
../diplomat/target/release/diplomat-tool kotlin ~/tmp/test_kt/
 cp target/release/libsomelib.dylib ~/tmp/test_kt/
```
Will create a kotlin project. I configure it with JVM 19 and add the following `Main.kt` to the module
```kt
package dev.gigapixel.somelib

object Main {

    @JvmStatic
    fun main(args: Array<String>) {
        val someStruct = SomeStruct.create(634, 772)
        println(someStruct.doSomething(3))
    }
}
```
the running of which correctly prints `7`

### CODE REVIEW

File: tool/templates/kotlin/DiplomatStr.java.jinja
Comment (jcrist1) [inline tool/templates/kotlin/DiplomatStr.java.jinja:1]: would it make sense to package this and related fundamentals as a standalone dependency?

Comment (Manishearth) [reply tool/templates/kotlin/DiplomatStr.java.jinja:1]: we've not been doing that so far because we're trying to be self contained. not opposed, but i think for now it's fine to just package and duplicate

File: tool/src/lib.rs
Comment (Manishearth) [inline tool/src/lib.rs:100]: nit: probably best to have this be taken from the CLI, or a config, in the long run. fine for now

we need to come up with a principled model for backend-specific config. Right now we have this `library_config` that you can set which is fine to use.

**Review by Manishearth — COMMENTED** [review]
I think this is a good start. Not sure when I'll get time to review this, so I suggest you keep pushing to this.

In general our attitude towards backends we do not maintain is "you can maintain them, we will try to fix them where possible but we may break them / disable tests / whatever if needed", so we don't *have* to review this fully, but I'd like to review to some extent.

I probably don't want to review this until there are some tests showing the final generated code. You can just hook into `feature_tests` if it works, or write some local insta tests.

**Comment by jcrist1** [comment]
> I probably don't want to review this until there are some tests showing the final generated code. You can just hook into `feature_tests` if it works, or write some local insta tests.

Thanks. That makes sense. Mostly just needed to know what's the best way to get incremental progress. I can take the informal tests I've implemented and integrate them, as well as adding attribute disabling for the functionality that doesn't work. Will try and get it in today still but realistically won't be till next week.

**Comment by Manishearth** [comment]
Yeah if you're writing Java specific tests put them in a separate folder somewhere, but in the long run we can integrate with feature_tests

File: feature_tests/kotlin/somelib/build.gradle.kts
Comment (jcrist1) [inline feature_tests/kotlin/somelib/build.gradle.kts:1]: I had to manually copy these files from the `tmp` directory (see above comment)

File: feature_tests/kotlin/somelib/src/main/java/dev/gigapixel/somelib/DiplomatStr.java
Comment (jcrist1) [inline feature_tests/kotlin/somelib/src/main/java/dev/gigapixel/somelib/DiplomatStr.java:7]: This is not used yet, but will be need to implement string slice compatibility, and is a prototype for struct types via jna that I'd like to keep around

File: feature_tests/kt-conf.toml
Comment (jcrist1) [inline feature_tests/kt-conf.toml:2]: I would prefer that this is picked up from the tool because it's user to load the dylib, and it would be convenient if it were the same as the generate lib file. But I couldn't figure it out

File: Makefile.toml
Comment (jcrist1) [reply Makefile.toml:456]: but I got the same error trying to run the dart gen so 🤷

File: feature_tests/kotlin/somelib/src/main/kotlin/dev/gigapixel/somelib/Lib.kt
Comment (jcrist1) [inline feature_tests/kotlin/somelib/src/main/kotlin/dev/gigapixel/somelib/Lib.kt:3]: top level cleaner for the whole library

File: feature_tests/kotlin/somelib/src/test/kotlin/dev/gigapixe/somelib/OtherOpaqueTest.kt
Comment (jcrist1) [inline feature_tests/kotlin/somelib/src/test/kotlin/dev/gigapixe/somelib/OtherOpaqueTest.kt:10]: it is a very silly test, but still it works.

File: feature_tests/src/structs.rs
Comment (jcrist1) [inline feature_tests/src/structs.rs:11]: This is just temporary until I get support for other return types. Then I can just use the existing structs and not pollute this file.

File: tool/Cargo.toml
Comment (jcrist1) [inline tool/Cargo.toml:35]: I like using tests to debug and develop the code gen and it helps to have these crates. Happy to remove them once I have the full functionality or before a possible merge, but I would like to keep them while iterative development is happening.

File: tool/src/kotlin/formatter.rs
Comment (jcrist1) [inline tool/src/kotlin/formatter.rs:15]: Will probably need to expand these.

File: tool/templates/kotlin/DiplomatStr.java.jinja
Comment (jcrist1) [reply tool/templates/kotlin/DiplomatStr.java.jinja:1]: 👍

File: tool/src/lib.rs
Comment (jcrist1) [reply tool/src/lib.rs]: agreed. Didn't really know how to pick those up. I'd like it if the actual library name could be picked up from Cargo.toml (at least as a default), but couldn't figure that out yet.

**Comment by jcrist1** [comment]
@Manishearth I added some codegen to the features_tests (while disabling _all_ of the existing examples  for kotlin 😅 and adding one new struct example). I'm not entirely sure how to plug them into the existing ones, as I couldn't get the duckscript file for generation to quite work. the generation outputed to a `tmp` directory successfully but copying to the new `kotlin` directory for `feature_tests` failed, so I copied them over manually. The test successfully loads the library and yields the correct construction and function call. I've left a bunch of preliminary comments on existing issues that will get resolved once later functionality is implemented.

Finally thanks for the attention you've already given. While I can only work (mostly) on the weekends on this, I don't expect feedback on the weekends, and am be grateful for it whenever you have time 😄.

Thanks to you and your co-maintainers for the cool library!

**Comment by jcrist1** [comment]
Finally I opened the PR up for review. I don't know what y'all's process. I'm fine putting this back to draft state if y'all want a more feature complete PR.

File: Makefile.toml
Comment (Manishearth) [reply Makefile.toml:456]: huh, i haven't seen that before. worth debugging

Comment (robertbastian) [reply Makefile.toml:456]: https://github.com/rust-diplomat/diplomat/issues/440, workaround is to install an older `cargo-make`

**Comment by Manishearth** [comment]
Our process is typically to just review this, but it is somewhat low priority for me atm so it might take some time (feel free to keep pushing here). Given that we are not going to be maintaining the backend my plan is to do a basic review, not a thorough one, and land it to let you keep iterating.

**Comment by jcrist1** [comment]
Merged changes to support lifetimes and borrows in opaque types

File: Makefile.toml
Comment (jcrist1) [reply Makefile.toml:456]: installed v0.36.13 and it seems to work

File: tool/templates/kotlin/opaqueReturn.kt.jinja
Comment (jcrist1) [inline tool/templates/kotlin/opaqueReturn.kt.jinja:9]: I originally thought about having the cleaner be called in the class on instantiation, but read somewhere that it shouldn't be called there. That it should only be registered after the class is instantiated, but I might have imagined that.

**Comment by jcrist1** [comment]
I now added support for slice types, excluding char, because as I understand it a char is a represented as a short in the JVM while it is a u32 in rust, and what exactly is happening on for codepoints above `UxFFFF` is a bit unclear for me. There's also some simplification of the method generation code as I no longer need a separate method for companion vs class methods

**Comment by jcrist1** [comment]
Merged main and noticed there are a bunch of cool new features. Had to comment out some feature tests that I'll uncomment once I've figured out a solution to the new functionalities

**Comment by jcrist1** [comment]
Added owned slices. Have to admit it makes me feel queasy, so I created a dedicated wrapper type. It feels a little contradictory because the owned slices are passed around to avoid allocations and memcopies, but I'm now paying for that with an atomic check.

**Comment by Manishearth** [comment]
cc @robertbastian on the design of owned slices

File: feature_tests/kotlin/somelib/src/main/kotlin/dev/gigapixel/somelib/Float64Vec.kt
Comment (robertbastian) [inline feature_tests/kotlin/somelib/src/main/kotlin/dev/gigapixel/somelib/Float64Vec.kt:136]: An owned slice should be a `DoubleArray` as well. In GC languages, owned and borrowed slices look the same, because they both always need to be reallocated into the Rust heap.

**Comment by Manishearth** [comment]
My general vibe on merging this is that I'd _prefer_ to merge this once there is struct/enum support. However I can understand if that is not coming any time soon.

Handling for borrows would also be ideal but that is more complicated.

**Comment by jcrist1** [comment]
> My general vibe on merging this is that I'd _prefer_ to merge this once there is struct/enum support. However I can understand if that is not coming any time soon.
> 
> Handling for borrows would also be ideal but that is more complicated.

I'm starting on structs now. I've got a basic working example of a struct in Kotlin loading from a dylib, just need to extract that into a template and add the parsing. Should be done sometime this weekend. Still need a bit to figure out enums, but it will probably be the following weekend. I think I have working borrows for opaques already, and expect I can use the same or similar for structs (with help from the existing dart backend code).

**Comment by jcrist1** [comment]
>  I think I have working borrows for opaques already, and expect I can use the same or similar for structs (with help from the existing dart backend code).

My original implementation missed a lot of details which I'm now fixing. I'll see if I can get them in and then push today.

**Comment by jcrist1** [comment]
Okay I pushed the struct implementation. There's a some code duplication that I'd like to DRY out. I'll have a look at that and  the enums this weekend, as well as adding some kotlin tests for the new structs.

File: feature_tests/kotlin/somelib/src/main/kotlin/dev/gigapixel/somelib/Float64Vec.kt
Comment (jcrist1) [reply feature_tests/kotlin/somelib/src/main/kotlin/dev/gigapixel/somelib/Float64Vec.kt]: Okay, owned slices are now just converted to kotlin types and the corresponding native memory is immediately freed.

**Comment by jcrist1** [comment]
@Manishearth I now have enums and structs. I also made owned slices just convert to kotlin types and free the associated memory, and introduced a test that tries to check for no memory leak. I sill feel like there could be a lot more clean up, but am worried about letting the perfect be the enemy of the good.
 My personal priority for follow ups, if/when this gets through review:
* fallible returns (I've commented out one test still as the structure in question had very many useful tests and I couldn't disable a single function).
* iterators (only now saw that you support these which looks cool!)
* start looking at a Panama backend
* mutable references and concurrency protection

The first two points I can continue pushing to this branch, but I would prefer to merge this before working on a new backend or mutable references.

Finally I was curious how enum variants with data are supported in FFI, and it seems that they [are possible](https://doc.rust-lang.org/reference/type-layout.html#combining-primitive-representations-of-enums-with-fields-and-reprc) if I'm reading it right, but the details aren't clear to me. Is it possible and are there plans to support those in Diplomat?

**Review by robertbastian — APPROVED** [review]
Looks good!

**Comment by jcrist1** [comment]
Okay @Manishearth I added enum returns but also got inspired and added nullables as best I could.

**Comment by Manishearth** [comment]
> Finally I was curious how enum variants with data are supported in FFI, and it seems that they [are possible](https://doc.rust-lang.org/reference/type-layout.html#combining-primitive-representations-of-enums-with-fields-and-reprc) if I'm reading it right, but the details aren't clear to me. Is it possible and are there plans to support those in Diplomat?

Well, ultimately whether or not rustc has a primitive representation for these is irrelevant since we can always codegen our own no matter what.

But yes, it's possible. Currently out of scope for Diplomat because for most languages there's no clear idiomatic equivalent. Diplomat tends to try and support features found in most languages, and includes features found in only some languages _if_ it's an easy extension (e.g. "marking a type as an iterator" is easy and doesn't require major typesystem changes). Dataful enums are probably out of scope for now. Maybe in the long run.

File: feature_tests/kt-conf.toml
Comment (Manishearth) [inline feature_tests/kt-conf.toml:1]: nit: please use something like `dev.diplomattest`, this feels like a third party package

(I know the domain doesn't exist)

File: Makefile.toml
Comment (Manishearth) [inline Makefile.toml:325]: nitL these tasks should be made a part of the corresponding gen/test tasks

(gen, gen-feature, test, test-feature)

File: feature_tests/kt-conf.toml
Comment (Manishearth) [reply feature_tests/kt-conf.toml:1]: also let's call this diplomat-kt-conf.toml

File: feature_tests/src/slices.rs
Comment (Manishearth) [inline feature_tests/src/slices.rs:124]: issue: please reenable this test, you can keep it disabled for kotlin

**Comment by Manishearth** [comment]
Make sure to fix CI

File: feature_tests/src/slices.rs
Comment (Manishearth) [inline feature_tests/src/slices.rs:41]: nit: this may need a `#[diplomat::skip_if_ast]` for CI to pass.

File: feature_tests/src/structs.rs
Comment (Manishearth) [inline feature_tests/src/structs.rs:85]: nit: everything returning opaque references will need  `#[diplomat::skip_if_ast]` for CI to pass.

File: feature_tests/src/structs.rs
Comment (Manishearth) [inline feature_tests/src/structs.rs:89]: nit: just allow on the module

File: feature_tests/src/structs.rs
Comment (Manishearth) [inline feature_tests/src/structs.rs:13]: nit: let's give this a more descriptive name for its purpose, maybe MutexedOpaque?

**Review by Manishearth — COMMENTED** [review]
Partial review, still need to review actual code

File: Makefile.toml
Comment (jcrist1) [reply Makefile.toml:325]: I added it, but while on the issue of generation. I made diplomat generate the whole kotlin project, so when I run the duckscript it deletes the whole kotlin feature project directory including the tests. Consequently I have to check out the test files before running the tests. Looking now at the other languages it seems like the project files are out of scope. I have to admit I prefer the idea of a tool that will spit out a ready library that's all packaged, but understand if that's not how y'all want to do it. If I can keep it generating the whole project, then I'd like to modify the duckscript so that it copies over a configured subdirectory (that defaults to the target directory for generation)  from the generated files in `tmp`.

Comment (Manishearth) [reply Makefile.toml:325]: That would be fine if you document it well enough.

Comment (jcrist1) [reply Makefile.toml:327]: Great! I added a new argument to the duckscript helper at the end with several explanatory comments. I also mention that the kotlin gen task only copies over the main source files and not the generated project files for now even though the backend creates them.

File: support/functions.ds
Comment (Manishearth) [inline support/functions.ds]: nit: document how the optional subdirectory on a separate line

File: support/functions.ds
Comment (Manishearth) [inline support/functions.ds:13]: nit: perhaps call this preserved_subdirectory?

File: tool/src/kotlin/mod.rs
Comment (Manishearth) [inline tool/src/kotlin/mod.rs:1097]: suggestion: you can use insta tests for these things. You can see how tool/src/js does it with the snapshots, and diplomat_core uses insta a lot.

File: feature_tests/kotlin/somelib/src/main/kotlin/dev/diplomattest/somelib/NestedBorrowedFields.kt
Comment (Manishearth) [inline feature_tests/kotlin/somelib/src/main/kotlin/dev/diplomattest/somelib/NestedBorrowedFields.kt:31]: issue: the borrow edges aren't getting passed down, I think? Is borrowing a WIP?

(that's fine if it is, just want to know)

**Review by Manishearth — APPROVED** [review]
overall looks good! I'm not sure what's going on with the borrowing


(I'm also curious: how confusing/easy was it to figure out the borrowing APIs here? Do you have questions about them?)

File: feature_tests/src/structs.rs
Comment (jcrist1) [reply feature_tests/src/structs.rs]: I called this OpaqueMutexedString

File: feature_tests/kotlin/somelib/src/main/kotlin/dev/diplomattest/somelib/NestedBorrowedFields.kt
Comment (jcrist1) [reply feature_tests/kotlin/somelib/src/main/kotlin/dev/diplomattest/somelib/NestedBorrowedFields.kt:31]: hmm I thought I had that. But I was just looking at the `Foo.extractFromBounds` (But of course would appreciate a check on the lifetimes in method params)... I think I understand now. I only handled borrows for method params but not for struct fields. This was just an oversight on my part.

Comment (jcrist1) [reply feature_tests/kotlin/somelib/src/main/kotlin/dev/diplomattest/somelib/NestedBorrowedFields.kt:31]: I'll work on the struct field borrows still

**Comment by jcrist1** [comment]
@Manishearth Thanks for the review. I've got one issue with the CI that I can't quite figure out. It just errors 
```
 +.. js:class:: Utf16Wrap
+
+    .. js:method:: borrow_cont()
+
+    .. js:method:: owned()
Error: Process completed with exit code 1.
```
Does that mean the `Utf16Wrap` is causing some issue with the JS backend? I can add a disable for js for that struct.

Regarding lifetimes: I thought it was reasonable to figure out the borrows for method params combined with looking at the dart code. I think it was just developer error that I didn't think about the fields 😅. I don't know what would have helped me spot that. I can kind of imagine having an enum describing the kind of borrows that a type will have, and handling all of them, but I can't see where that would come up in the backend code because field borrows and method borrows are in completely different generation places

I guess the best would be to develop a checklist for the [backend documentation](https://rust-diplomat.github.io/book/developer.html). It would be cool to incorporate that checklist in a PR template for new backends or issues. I could, for instance, expand the checklist I have in the PR description, add links to the HIR types and open up a PR for the docs if y'all're interested.

File: support/functions.ds
Comment (jcrist1) [reply support/functions.ds]: I added a much more detailed explanation of the optional subdirectory

File: tool/src/kotlin/mod.rs
Comment (jcrist1) [reply tool/src/kotlin/mod.rs:1104]: It might take me a while to get to know insta. But I'll have a look.

**Comment by jcrist1** [comment]
@Manishearth I think I have the borrows, but admit that it feels thorny. I added simple methods for the borrowing structs `BorrowedFields`, `BorrowedFieldsWithBounds` and `NestedBorrowedFields`, to see that the attached lifetime edges should correspond when I e.g. create a `NestedBorrowedFields`, compared to if I just create the fields individually, and I _think_ it is consistent now, but don't feel like I can _prove_ it.

I'm going to start looking at the `insta` now in place of the tests that I have, but don't really know how long that will take, so I pushed the changes in case you want to look at the borrows already.

**Comment by jcrist1** [comment]
It didn't take long, but getting something with the generate files like in the js-gen will take a while.

**Comment by jcrist1** [comment]
Okay insta now generates snapshot files instead.

**Comment by jcrist1** [comment]
```
+    internal val xEdges: List<Any>,
     internal val zEdges: List<Any>,
-    internal val yEdges: List<Any>,
-    internal val xEdges: List<Any>
+    internal val yEdges: List<Any>
```
The `gen` step is failing I think because the lifetime edge order is non-deterministic. I'll try to add a fix

**Comment by Manishearth** [comment]
> lifetime edge order is non-deterministic

This surprises me, the edges are ordered according to how they show up in the code. I guess your code is sticking them in some non deterministic map?

**Comment by jcrist1** [comment]
Okay I have made changes to address all of the ci errors I'm seeing. I can't get the feature test for dart to run locally but I think it's just my dart setup.

**Review by Manishearth — APPROVED** [review]

**Comment by Manishearth** [comment]
I'm not convinced the borrowing is correct but I don't expect correct borrowing for the first pass anyway.

**Comment by Manishearth** [comment]
(and I'm also okay with incorrect borrowing in a not-fully-officially-supported backend)

**Comment by Manishearth** [comment]
For dart I think you'll need to rename the method that you've called `str()`

**Comment by jcrist1** [comment]
Alright no more extra `str` method in dart.

**Review by Manishearth — APPROVED** [review]

**Comment by Manishearth** [comment]
Sweet, thanks!

**Pull request merged**
